src/lib/optimizer/abstract_syntax_tree/abstract_ast_node.cpp

```diff
@@ -181,37 +224,56 @@ std::vector<ColumnID> AbstractASTNode::get_output_column_ids_for_table(const std
     return get_output_column_ids();
   }
 
-  return _left_child->get_output_column_ids_for_table(table_name);
+  return left_child()->get_output_column_ids_for_table(table_name);
 }
 
 void AbstractASTNode::remove_from_tree() {
-  Assert(!_right_child, "Can't remove a node with two children");
+  Assert(left_child() && !right_child(), "Can only rermove nodes that only have a left child");
 
-  auto parent = _parent.lock();
+  /**
+   * Back up parents and in which child side they hold this node
+   */
+  auto parents = this->parents();
+  auto child_sides = this->get_child_sides();
 
-  if (parent) {
-    parent->set_left_child(_left_child);  // Note: It's totally fine for _left_child to be a nullptr
-  } else if (_left_child) {
-    _left_child->clear_parent();
+  // This will call get_child_side() again, but let's avoid code duplication
+  clear_parents();
+
+  /**
+   * Hold left_child ptr in extra variable and untie it from this node
+   */
+  auto left_child = this->left_child();
+  set_left_child(nullptr);
+
+  /**
+   * Tie this nodes previous parents with this nodes previous left child
+   */
+  if (left_child) {
+    for (size_t parent_idx = 0; parent_idx < parents.size(); ++parent_idx) {
+      parents[parent_idx]->set_child(child_sides[parent_idx], left_child);
+    }
   }
 }
 
 void AbstractASTNode::replace_in_tree(const std::shared_ptr<AbstractASTNode>& node_to_replace) {
-  Assert(!_left_child && !_right_child && !parent(),
-         "Can't put a Node that's already part of a tree into another tree. Call remove_from_tree() first");
+  DebugAssert(_parents.empty(), "Node can't have parents");
+  DebugAssert(!left_child() && !right_child(), "Node can't have children");
 
-  set_left_child(node_to_replace->left_child());
-  set_right_child(node_to_replace->right_child());
+  const auto parents = node_to_replace->parents();
+  const auto child_sides = node_to_replace->get_child_sides();
+  auto left_child = node_to_replace->left_child();
+  auto right_child = node_to_replace->right_child();
 
-  auto parent = node_to_replace->parent();
-  if (parent) {
-    if (parent->left_child() == node_to_replace) {
-      parent->set_left_child(shared_from_this());
-    } else {
-      Assert(parent->right_child() == shared_from_this(), "Invalid binary tree");
-      parent->set_right_child(shared_from_this());
-    }
+  node_to_replace->clear_parents();
+  node_to_replace->set_left_child(nullptr);
```

Comment:
There should be a comment here that this is done so that the reference counters are correct.

Reply:
not keeping any instances for ref counts any more. done.